src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/CngKey.StandardProperties.cs

```diff
@@ -40,14 +49,22 @@ public CngAlgorithm Algorithm
         ///     Name of the algorithm this key can be used with
         /// </summary>
         public CngAlgorithmGroup? AlgorithmGroup
-
         {
             get
             {
-                string? algorithmGroup = _keyHandle.GetPropertyAsString(KeyPropertyName.AlgorithmGroup, CngPropertyOptions.None);
-                if (algorithmGroup == null)
-                    return null;
-                return new CngAlgorithmGroup(algorithmGroup);
+                if (!_hasCachedAlgorithmGroup || _keyHandle.IsClosed)
+                {
+                    string? algorithmGroup = _keyHandle.GetPropertyAsString(KeyPropertyName.AlgorithmGroup, CngPropertyOptions.None);
+
+                    if (algorithmGroup is not null)
+                    {
+                        _cachedAlgorithmGroup = new CngAlgorithmGroup(algorithmGroup);
+                    }
+
+                    _hasCachedAlgorithmGroup = true;
+                }
+
+                return _cachedAlgorithmGroup;
```

Comment:
Is CngKey supported for multithreaded reads? If so, this needs some volatile love.

Reply:
Hm, reads, yeah. Good point.